### PR TITLE
Extend French Peppol exports for CTC and participant routing

### PR DESCRIPTION
**Pull request by djukicmilica**
## What & why

Extend the French Peppol BIS 3.0 format to support Extended CTC France scenarios and France-specific invoice metadata required for compliant routing and export.

This change:
- selects the `EXTENDED-CTC-FR` customization when an invoice contains multiple distinct order references, shipment references, or shipment posting dates, while retaining the basic profile for repeated references;
- adds the Extended CTC line-level order, shipment, and delivery-date references;
- exports typed French regulatory comments from posted sales invoices and credit memos as UBL header notes;
- sets the French billing mode (`B1`, `S1`, or `M1`) from the document line types;
- allows service-specific participant identifiers and French identifier schemes to override customer and company endpoint defaults;
- adds scheme `0225` for French CTC routing and scheme `9957` for French VAT endpoint fallbacks when explicit endpoints are missing;
- adds French sales and service validation implementations while retaining the standard Peppol handling for other supported document types;
- removes Purchase Invoice from the outbound source document types registered for the French Peppol format; received Peppol invoices and credit notes continue to use the standard import parser.

## Linked work

[AB#642536](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/642536)

An approved GitHub issue still needs to be linked before this PR is ready for review.

## How I validated this

- [ ] I read the full diff and it contains only changes I intended.
- [ ] I built the affected app(s) locally with no new analyzer warnings.
- [ ] I ran the change in Business Central and confirmed it behaves as expected.
- [x] I added or updated tests for the new behavior, or explained below why none are needed.

**What I tested and the outcome**

Added focused tests in `PEPPOLBIS30XMLTests.Codeunit.al` covering:
- Extended CTC selection for multiple order references and multiple shipment references;
- retention of the basic profile for repeated references;
- Extended CTC line-level order, shipment, and delivery-date references;
- export of French regulatory comments as UBL notes;
- company and customer endpoint resolution through service participants;
- scheme `0225` routing identifiers and scheme `9957` French VAT fallbacks;
- seller and buyer VAT fallback behavior when explicit electronic addresses are absent;
- rejection of French VAT fallback for non-French buyers;
- French validation behavior for sales and service documents.

Current required GitHub checks pass, including Pull Request Build and Work Item Validation.

## Risk & compatibility

Endpoint selection precedence changes for the French Peppol format: a service participant identifier now takes precedence over values stored on the company or customer, and French VAT registration numbers are used as fallback endpoints. Existing explicit participant or electronic-address configuration remains supported.

The French format no longer registers Purchase Invoice as an outbound source document type. Received Peppol invoices and credit notes continue to be parsed into temporary purchase documents through the standard Peppol importer.

The added fields and French Peppol format enum values are additive. No data migration is required.

### CODE REVIEW

File: src/Apps/FR/EDocument_FR/EReportingFR/app/src/Core/FRRegulatoryCommentType.Enum.al [hunk 1]
```diff
@@ -0,0 +1,27 @@
+// ------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+// ------------------------------------------------------------------------------------------------
+namespace Microsoft.eServices.EDocument.Formats;
+
+enum 10971 "FR Regulatory Comment Type"
+{
+    Extensible = false;
```
Comment (nicolassaleron) [inline src/Apps/FR/EDocument_FR/EReportingFR/app/src/Core/FRRegulatoryCommentType.Enum.al]: There are actually many more codes. Would it be possible to implement all of them? If not, could this enum at least be made extensible?

See the complete list here: https://service.unece.org/trade/untdid/d24a/tred/tred4451.htm

Comment (djukicmilica) [reply src/Apps/FR/EDocument_FR/EReportingFR/app/src/Core/FRRegulatoryCommentType.Enum.al]: @nicolassaleron thanks for the verification!
Yes, will we take into account the list.

File: src/Apps/FR/EDocument_FR/EReportingFR/app/src/Formats/PEPPOL/PeppolBIS30FRFormat.Codeunit.al [hunk 2]
```diff
@@ -93,6 +98,138 @@ codeunit 10977 "Peppol BIS 3.0 FR Format" implements "E-Document"
...
+                exit;
+        end;
+
+        SalesCommentLine.SetRange("Document Type", DocumentType);
+        SalesCommentLine.SetRange("No.", DocumentNo);
+        SalesCommentLine.SetRange("Document Line No.", 0);
+        SalesCommentLine.SetFilter("FR Regulatory Comment Type", '<>%1', SalesCommentLine."FR Regulatory Comment Type"::None);
+        if SalesCommentLine.FindSet() then
+            repeat
+                NoteElement := XmlElement.Create('Note', CbcNamespaceTok, SalesCommentLine.Comment);
```
Comment (nicolassaleron) [inline src/Apps/FR/EDocument_FR/EReportingFR/app/src/Formats/PEPPOL/PeppolBIS30FRFormat.Codeunit.al]: Hello,

First of all, sorry if I’m interfering with your early development process.

The purpose of tagging comment lines with the **FR Regulatory Comment Type** is to automatically prefix the comment with the corresponding comment type.

For example, I would expect the following **Sales Comment Line** content:

| Comment | FR Regulatory Comment Type |
|--|--|
| Tout retard de paiement engendre une pénalité exigible à compter de la date d'échéance, calculée sur la base de trois fois le taux d'intérêt légal. | PMD |

to be exported as the following XML element:

`<cbc:Note>#PMD#Tout retard de paiement engendre une pénalité exigible à compter de la date d'échéance, calculée sur la base de trois fois le taux d'intérêt légal.</cbc:Note>`

If users are required to manually prefix the comment itself with `#PMD#` in the **Comment** field, then strongly typing the comment with a dedicated type seems unnecessary. In that case, a simple checkbox indicating that the comment should be exported as a regulatory comment would achieve the same result.

Comment (djukicmilica) [reply src/Apps/FR/EDocument_FR/EReportingFR/app/src/Formats/PEPPOL/PeppolBIS30FRFormat.Codeunit.al]: Hi @nicolassaleron,

Thanks for your inputs.

The selected FR Regulatory Comment Type is automatically added as a prefix during export. Users enter only the comment text; for example, a comment typed as PMD is exported as <cbc:Note>#PMD#Comment text</cbc:Note>. Manual prefixing is not required.

I updated tooltips for better understanding for users.

File: src/Apps/FR/EDocument_FR/EReportingFR/app/src/Extensions/FRServiceParticipants.PageExt.al [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+pageextension 10980 "FR Service Participants" extends "Service Participants"
+{
+    layout
+    {
+        addafter("Participant Identifier")
+        {
+            field("FR Identifier Scheme"; Rec."FR Identifier Scheme")
+            {
+                ApplicationArea = All;
+            }
```
[the remark this answers is not in the document]
Comment (djukicmilica) [reply src/Apps/FR/EDocument_FR/EReportingFR/app/src/Extensions/FRServiceParticipants.PageExt.al:18]: it is on table

File: src/Apps/FR/EDocument_FR/EReportingFR/app/src/Extensions/FRSalesCommentSheet.PageExt.al [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+pageextension 10975 "FR Sales Comment Sheet" extends "Sales Comment Sheet"
+{
+    layout
+    {
+        addafter(Comment)
+        {
+            field("FR Regulatory Comment Type"; Rec."FR Regulatory Comment Type")
+            {
+                ApplicationArea = Comments;
+            }
```
[the remark this answers is not in the document]
Comment (djukicmilica) [reply src/Apps/FR/EDocument_FR/EReportingFR/app/src/Extensions/FRSalesCommentSheet.PageExt.al:18]: it is on table

File: src/Apps/FR/EDocument_FR/EReportingFR/app/src/Formats/PEPPOL/PEPPOL30FormatFR.EnumExt.al [hunk 5]
```diff
@@ -0,0 +1,23 @@
+// ------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+// ------------------------------------------------------------------------------------------------
+namespace Microsoft.eServices.EDocument.Formats;
+
+using Microsoft.Peppol;
+
+enumextension 10995 "PEPPOL 3.0 Format FR" extends "PEPPOL 3.0 Format"
```
[the remark this answers is not in the document]
Comment (djukicmilica) [reply src/Apps/FR/EDocument_FR/EReportingFR/app/src/Formats/PEPPOL/PEPPOL30FormatFR.EnumExt.al:9]: The new workitem has been created for this as it is a part of peppol app
https://dynamicssmb2.visualstudio.com/Dynamics%20SMB/_workitems/edit/646302

**Comment by alexei-dobriansky** [comment]
## Agentic PR Review - Round 1

# **Recommendation: Accept**

<details>
<summary><strong>What this PR does</strong></summary>

This Slice extends the French Peppol and Factur-X export paths for Extended CTC France, French routing schemes, service participant endpoints, regulatory comments, billing mode, and French sales/service validation. The diff is coherent for that feature goal: endpoint precedence is checked before export, the new XML mutations are limited to the French format, and the Extended CTC line references are added only when the posted invoice has multiple distinct order references, shipment references, or delivery dates. The tests exercise the main routing, validation, VAT fallback, regulatory comment, and Extended CTC cases, and the PR checks show the affected FR apps and tests build successfully.
</details>

<details>
<summary><strong>Suggestions</strong></summary>

*None.*
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The regression surface is mainly French e-document export, especially endpoint selection, validation errors, and invoice XML content used for compliant routing. The change does not depend on a BaseApp publisher or event-ordering change; the new events are local extension points. The purchase-invoice outbound registration removal is called out in the PR description, while the parser path remains in place.

**Necessity:** The work item is a Slice for required French e-invoicing improvements, and the PR description plus tests give enough feature evidence. The scope is large but focused on the French Peppol/CTC increment and includes targeted coverage for the new behavior.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=2 system=github pr=9639 round=1 by=alexei-dobriansky at=2026-08-17T13:18:11Z lastSha=07cee7d1fab691fc6b090ee406335514f0b20d87 reviewKey=92c0c5eb8d9103649df76bda1acbad1b188724f203ed7cef293f62123ff19874 suggestions=

**Comment by alexei-dobriansky** [comment]
## Agentic PR Review - Round 2

# **Recommendation: Accept**

<details>
<summary><strong>What this PR does</strong></summary>

This Slice extends the French Peppol and Factur-X export paths for Extended CTC France, French routing identifiers, service participant endpoints, regulatory comments, billing mode, and French sales/service validation. The new commit since round 1 only adjusts test fixtures: it gives the Peppol and Factur-X test posting setup separate codes and populates service document ship-to fields, which matches the test-isolation goal and does not change product behavior.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|----|-------|--------|-----------------|
| — | No previous suggestions | — | — |
</details>

<details>
<summary><strong>New observations (commits since round 1)</strong></summary>

*None - the new commit only isolates test setup and does not add a new issue in the net PR diff.*
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The product regression surface remains French e-document export, especially endpoint selection, validation, billing mode, and XML content used for routing. The round-2 commit is limited to tests and reduces fixture collision risk by avoiding shared posting setup codes across test codeunits.

**Necessity:** The work item is a Slice for required French e-invoicing improvements. The current PR scope is still coherent for that feature, and the added test-isolation commit is necessary to keep the new coverage stable.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=2 system=github pr=9639 round=2 by=alexei-dobriansky at=2026-08-18T13:14:51Z lastSha=f99f6d9c2c5d8d1f2daaae57baed5a0f5a9e8b42 reviewKey=77cb161dfec0745382a359a628d1004a78426ac8d71384eb2b41460b5cdc13e0 suggestions= parentRound=1

File: src/Apps/FR/EDocument_FR/EReportingFR/app/src/Core/CIIXMLBuilder.Codeunit.al [hunk 6]
```diff
@@ -321,6 +317,14 @@ codeunit 10978 "CII XML Builder"
         PartyElement.Add(TaxRegElement);
     end;
 
+    local procedure GetVATRegistrationNoWithCountryPrefix(VATRegistrationNo: Text; CountryCode: Text): Text
+    begin
+        if (StrLen(VATRegistrationNo) >= 2) and (DelChr(CopyStr(VATRegistrationNo, 1, 2), '=', 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz') = '') then
```
Comment (dcenic) [inline src/Apps/FR/EDocument_FR/EReportingFR/app/src/Core/CIIXMLBuilder.Codeunit.al]: If this is not on a hot-path performance-wise, consider using RegEx codeunit and calling
Regex.Replace(YourString, '[A-Za-z]', '');
It is less error-prone and more readable (no need to spell all characters out), but involves invoking RegEx wrapper (DelChr is slightly faster).

File: src/Apps/FR/EDocument_FR/EReportingFR/app/src/Core/CIIXMLBuilder.Codeunit.al [hunk 7]
```diff
@@ -506,58 +558,103 @@ codeunit 10978 "CII XML Builder"
...
 
+    [IntegrationEvent(false, false)]
+    local procedure OnBeforeAddExchangedDocumentContext(var RootElement: XmlElement; var SourceDocumentLines: RecordRef; var IsHandled: Boolean)
+    begin
+    end;
+
+    [IntegrationEvent(false, false)]
+    local procedure OnAfterAddExchangedDocumentContext(var RootElement: XmlElement; var SourceDocumentLines: RecordRef)
+    begin
+    end;
```
Comment (dcenic) [inline src/Apps/FR/EDocument_FR/EReportingFR/app/src/Core/CIIXMLBuilder.Codeunit.al:658]: Why do we need these IsHandled events, did someone explicitly request them?
E.g. the bot had a comment about the early exit after OnBefore (which makes OnAfter never to be raised).
I would not introduce these events unless they are needed, and I would try to avoid the overuse of IsHandled.
E-Documents have an interface, can it be handled through that?

File: src/Apps/FR/EDocument_FR/EReportingFR/app/src/Core/CIIXMLBuilder.Codeunit.al [hunk 8]
```diff
@@ -1047,9 +1164,11 @@ codeunit 10978 "CII XML Builder"
     var
         RsmNamespaceTok: Label 'urn:un:unece:uncefact:data:standard:CrossIndustryInvoice:100', Locked = true;
         RamNamespaceTok: Label 'urn:un:unece:uncefact:data:standard:ReusableAggregateBusinessInformationEntity:100', Locked = true;
+        QdtNamespaceTok: Label 'urn:un:unece:uncefact:data:standard:QualifiedDataType:100', Locked = true;
         UdtNamespaceTok: Label 'urn:un:unece:uncefact:data:standard:UnqualifiedDataType:100', Locked = true;
         FacturXProfileIdTok: Label 'urn:cen.eu:en16931:2017', Locked = true;
         RecoveryCostNoteTok: Label 'Indemnité forfaitaire pour frais de recouvrement en cas de retard de paiement : 40 €', Locked = true;
         LatePaymentPenaltyNoteTok: Label 'Taux des pénalités de retard : taux directeur (BCE) majoré de 10 points', Locked = true;
         EarlyPaymentDiscountNoteTok: Label 'Pas d''escompte pour paiement anticipé', Locked = true;
+        VATExemptionReasonLbl: Label 'Exempt from VAT', Locked = true;
```
Comment (dcenic) [inline src/Apps/FR/EDocument_FR/EReportingFR/app/src/Core/CIIXMLBuilder.Codeunit.al:1175]: This will be written in English to the XML? Is that OK, or should it be in French?

**Review by AndersLarsenMicrosoft — DISMISSED** [review]

**Comment by alexei-dobriansky** [comment]
## Agentic PR Review - Round 3

# **Recommendation: Accept**

<details>
<summary><strong>What this PR does</strong></summary>

This Slice extends the French Peppol and Factur-X export paths for Extended CTC France, routing identifiers, service participant endpoints, regulatory comments, billing mode, and French validation. The new commit since round 2 changes the Factur-X CII buyer VAT prefix check to use an anchored two-letter regex, adds edge-case tests for lowercase and incomplete prefixes, and tightens test isolation and regulatory-note assertions. The new CII logic still prefixes the country code only when the VAT value does not already start with two letters, so it matches the tested routing behavior without changing the surrounding XML flow.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|----|-------|--------|-----------------|
| — | No previous suggestions | — | — |
</details>

<details>
<summary><strong>New observations (commits since round 2)</strong></summary>

*None - the new commit is covered by targeted tests and I did not find a new issue in the changed spans that are also present in the net PR diff.*
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The product change affects Factur-X buyer VAT registration output for French CII XML. The regression surface is limited to the buyer tax registration value; the later XML construction still writes the same SpecifiedTaxRegistration node and scheme. The other changes are tests only. No BaseApp event publisher dependency is introduced by this round.

**Necessity:** The work item is a Slice for required French e-invoicing improvements. The round-3 change is justified because it locks down VAT prefix edge cases and improves the test proof for the feature without expanding the product scope.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=2 system=github pr=9639 round=3 by=alexei-dobriansky at=2026-08-19T19:10:23Z lastSha=c4026ee2affa053470ca445eab18194dc202372c reviewKey=52b4fd85d1db5bf3618d2d0aced85d6e181d9b1ef7ed6e1acde409b8c5523188 suggestions= parentRound=2

**Review by dcenic — APPROVED** [review]

File: src/Apps/FR/EDocument_FR/EReportingFR/app/src/Core/CIIXMLBuilder.Codeunit.al (hunk 8 above)
Comment (djukicmilica) [reply src/Apps/FR/EDocument_FR/EReportingFR/app/src/Core/CIIXMLBuilder.Codeunit.al:1175]: Thanks, I have another workitem where I will do the corrections

https://dynamicssmb2.visualstudio.com/Dynamics%20SMB/_workitems/edit/647217

File: src/Apps/FR/EDocument_FR/EReportingFR/app/src/Core/CIIXMLBuilder.Codeunit.al (hunk 7 above)
Comment (djukicmilica) [reply src/Apps/FR/EDocument_FR/EReportingFR/app/src/Core/CIIXMLBuilder.Codeunit.al:658]: Thanks, I have another workitem where I will do the corrections

https://dynamicssmb2.visualstudio.com/Dynamics%20SMB/_workitems/edit/647217

File: src/Apps/FR/EDocument_FR/EReportingFR/app/src/Core/CIIXMLBuilder.Codeunit.al [hunk 9]
```diff
@@ -453,49 +461,95 @@ codeunit 10978 "CII XML Builder"
...
-        AllocateInvoiceDiscountByVATKey(InvoiceDiscountAmount, LineBaseAmounts, AllocatedDiscountByKey);
-
         VATAggregationKeys := LineBaseAmounts.Keys();
+        foreach VATKey in VATAggregationKeys do begin
+            DiscountAmount := LineBaseAmounts.Get(VATKey) - LineNetBaseAmounts.Get(VATKey);
+            AllocatedDiscountByKey.Add(VATKey, DiscountAmount);
+            AllocatedDiscountTotal += DiscountAmount;
+        end;
+
+        if Round(AllocatedDiscountTotal - InvoiceDiscountAmount, 0.01) <> 0 then begin
```
Comment (AleksanderGladkov) [inline src/Apps/FR/EDocument_FR/EReportingFR/app/src/Core/CIIXMLBuilder.Codeunit.al:548]: should the rounding precision be taken from some setup?

Comment (djukicmilica) [reply src/Apps/FR/EDocument_FR/EReportingFR/app/src/Core/CIIXMLBuilder.Codeunit.al:548]: thanks, will check it out and update in the next PR

https://dynamicssmb2.visualstudio.com/Dynamics%20SMB/_workitems/edit/647217

**Review by AleksanderGladkov — APPROVED** [review]

File: src/Apps/FR/EDocument_FR/EReportingFR/app/src/Formats/PEPPOL/PeppolBIS30FRFormat.Codeunit.al [hunk 10]
```diff
@@ -217,36 +631,32 @@ codeunit 10977 "Peppol BIS 3.0 FR Format" implements "E-Document"
...
+        Customer.SetLoadFields("FR Electronic Address", "Registration Number", "VAT Registration No.");
         if not Customer.Get(CustomerNo) then
             exit(false);
 
-        ElecAddress := Customer."FR Electronic Address";
-        ElecAddressScheme := Customer."FR Elec. Address Scheme";
-        exit(ElecAddress <> '');
+        if not FREDocHelpers.GetBuyerElectronicAddress(Customer, ElecAddress) then
+            exit(false);
+        ElecAddressScheme := ElecAddressScheme::"0225";
```
Comment (nicolassaleron) [inline src/Apps/FR/EDocument_FR/EReportingFR/app/src/Formats/PEPPOL/PeppolBIS30FRFormat.Codeunit.al:645]: @djukicmilica Why forcing Scheme 0225? We got errors from PA (formerly PDP) about the scheme.
Indeed, if we setup SIREN or SIRET as a scheme on the customer card, then scheme 0225 is always sent instead of respecting the setup from the customer card.
